kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -284,6 +372,333 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
+          emptyList()
+        }
+      }
+      .plus(simpleName)
+
+  return "${packageName.replace("/", ".")}.${simpleNames.joinToString(".")}"
+}
+
+@Suppress("SameParameterValue")
+private fun String.substringAfterLast(vararg delimiters: Char): String {
+  val index = lastIndexOfAny(delimiters)
+  return if (index == -1) this else substring(index + 1, length)
+}
+
+private fun defaultPrimitiveValue(type: Type): Any? =
+    if (type is Class<*> && type.isPrimitive) {
+      when (type) {
+        Boolean::class.java -> false
+        Char::class.java -> 0.toChar()
+        Byte::class.java -> 0.toByte()
+        Short::class.java -> 0.toShort()
+        Int::class.java -> 0
+        Float::class.java -> 0f
+        Long::class.java -> 0L
+        Double::class.java -> 0.0
+        Void.TYPE -> throw IllegalStateException("Parameter with void type is illegal")
+        else -> throw UnsupportedOperationException("Unknown primitive: $type")
+      }
+    } else null
+
+private fun Class<*>.toKmClass(throwOnNotClass: Boolean): KmClass? {
+  val metadata = getAnnotation(KOTLIN_METADATA) ?: return null
+  val header = with(metadata) {
+    KotlinClassHeader(
+        kind = kind,
+        metadataVersion = metadataVersion,
+        bytecodeVersion = bytecodeVersion,
+        data1 = data1,
+        data2 = data2,
+        extraString = extraString,
+        packageName = packageName,
+        extraInt = extraInt
+    )
+  }
+  val classMetadata = KotlinClassMetadata.read(header)
+  if (classMetadata !is KotlinClassMetadata.Class) {
+    if (throwOnNotClass) {
+      throw IllegalStateException("Cannot serialize class with metadata $classMetadata")
+    } else {
+      return null
+    }
+  }
+
+  return classMetadata.toKmClass()
+}
+
+private class JvmSignatureSearcher(clazz: Class<*>) {
+
+  private val methodSignatureCache = mutableMapOf<String, Method>()
+  private val fieldSignatureCache = mutableMapOf<String, Field>()
+  private val declaredMethodsIterator by lazy(NONE) { clazz.declaredMethods.iterator() }
+  private val declaredFieldsIterator by lazy(NONE) { clazz.declaredFields.iterator() }
+  private val methodIterator by lazy(NONE) { clazz.methods.iterator() }
+  private val fieldIterator by lazy(NONE) { clazz.fields.iterator() }
+
+  fun findMethod(signature: JvmMethodSignature): Method? {
+    val signatureString = signature.asString()
+    val cached = methodSignatureCache[signatureString]
+    if (cached != null) return cached
+
+    while (declaredMethodsIterator.hasNext()) {
+      val next = declaredMethodsIterator.next()
+      val nextSignature = next.jvmMethodSignature
+      methodSignatureCache[nextSignature] = next
+      if (nextSignature == signatureString) {
+        return next
+      }
+    }
+
+    while (methodIterator.hasNext()) {
+      val next = methodIterator.next()
+      val nextSignature = next.jvmMethodSignature
+      methodSignatureCache[nextSignature] = next
```

Comment:
weird thing . . . declared methods and methods overlap.

Reply:
Cleaned up a bit in 5db8757